File: backend/src/atlas/modules/ai/domain/protected_draft_adjudication.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime


def _ids(*values: str) -> bool:
    return all(3 <= len(value.strip()) <= 256 for value in values)


def _digests(*values: str) -> bool:
    return all(
        len(value) == 64 and all(char in "0123456789abcdef" for char in value) for value in values
    )


@dataclass(frozen=True, slots=True)
class ProtectedDraftAdjudicationPolicySnapshot:
    policy_id: str
    schema_version: str
    version: int
    organization_id: str
    environment_id: str
    policy_version: str
    required_invocation_schema: str
    required_invocation_state: str
    required_draft_schema: str
    required_adjudicator_id: str
    required_adjudicator_attestor_id: str
    required_receipt_schema: str
    protected_vault_id: str
    validation_profile_digest: str
    prohibited_output_profile_digest: str
    browser_binding_key_digest: str
    classification_ceiling: str
    maximum_authentication_age_minutes: int
    maximum_summary_characters: int
    minimum_citation_count: int
    minimum_unknown_count: int
    retention_minutes: int
    signed_by: str
    signature_verified: bool
    issued_at: datetime
    expires_at: datetime
    canonical_digest: str

    def __post_init__(self) -> None:
        if (
            self.version != 1
            or not _ids(
                self.policy_id,
                self.schema_version,
                self.organization_id,
                self.environment_id,
                self.required_invocation_schema,
                self.required_invocation_state,
                self.required_draft_schema,
                self.required_adjudicator_id,
                self.required_receipt_schema,
                self.protected_vault_id,
            )
            or not _digests(
                self.validation_profile_digest,
                self.prohibited_output_profile_digest,
                self.browser_binding_key_digest,
                self.canonical_digest,
            )
            or not 1 <= self.maximum_authentication_age_minutes <= 60
            or not 100 <= self.maximum_summary_characters <= 100_000
            or not 0 <= self.minimum_citation_count <= 1_000
            or not 0 <= self.minimum_unknown_count <= 1_000
            or not 1 <= self.retention_minutes <= 1_440
            or self.issued_at.tzinfo is None
            or self.expires_at.tzinfo is None
            or not self.issued_at < self.expires_at
            or not self.signature_verified
        ):
            raise ValueError("Protected draft-adjudication policy is invalid")


@dataclass(frozen=True, slots=True)
class ProtectedDraftAdjudicationInstruction:
    adjudication_id: str
    invocation_id: str
    invocation_digest: str
    context_id: str
    context_digest: str
    context_package_digest: str
    draft_digest: str
    citation_set_digest: str
    output_safety_digest: str
    organization_id: str
    environment_id: str
    consumer_subject_digest: str
    adjudication_authorization_digest: str
    policy_id: str
    policy_digest: str
    validation_profile_digest: str
    prohibited_output_profile_digest: str
    minimum_citation_count: int
    minimum_unknown_count: int
    maximum_summary_characters: int
    requested_at: datetime
    expires_at: datetime


@dataclass(frozen=True, slots=True)
class ProtectedDraftAdjudicationReport:
    adjudication_id: str
    invocation_id: str
    draft_digest: str
    outcome: str
    check_codes: tuple[str, ...]
    citation_count: int
    unknown_count: int
    summary_character_count: int
    generated_at: datetime
    expires_at: datetime
    canonical_digest: str


@dataclass(frozen=True, slots=True)
class ProtectedDraftAdjudicationReceipt:
    adjudication_id: str
    schema_version: str
    version: int
    adjudicator_id: str
    attested_by: str
    invocation_id: str
    invocation_digest: str
    context_digest: str
    draft_digest: str
    adjudication_authorization_digest: str
    policy_digest: str
    protected_report_reference: str
    protected_report_digest: str
    report_digest: str
    check_set_digest: str
    citation_coverage_digest: str
    unknown_preservation_digest: str
    prohibited_output_digest: str
    check_count: int
    citation_count: int
    unknown_count: int
    outcome: str
    adjudicated_at: datetime
    expires_at: datetime
    schema_verified: bool
    citations_verified: bool
    unknowns_verified: bool
    prohibited_output_verified: bool
    no_model_used: bool
    protected_vault_write_verified: bool
    signature_verified: bool
    canonical_digest: str


@dataclass(frozen=True, slots=True)
class ProtectedDraftAdjudicationClaim:
    claim_id: str
    schema_version: str
    version: int
    adjudication_id: str
    invocation_id: str
    claimed_by_subject_digest: str
    browser_session_binding_digest: str
    request_binding_digest: str
    idempotency_digest: str
    organization_id: str
    environment_id: str
    claimed_at: datetime
    canonical_digest: str


@dataclass(frozen=True, slots=True)
class ProtectedDraftAdjudicationRecord:
    adjudication_id: str
    schema_version: str
    version: int
    claim_id: str
    invocation_id: str
    invocation_digest: str
    context_id: str
    context_digest: str
    organization_id: str
    environment_id: str
    classification: str
    consumer_subject_digest: str
    browser_session_binding_digest: str
    adjudication_policy_id: str
    adjudication_policy_digest: str
    adjudication_policy_version: str
    adjudicator_id: str
    adjudication_receipt_digest: str
    adjudication_authorization_digest: str
    draft_digest: str
    protected_report_reference: str
    protected_report_digest: str
    report_digest: str
    check_set_digest: str
    citation_coverage_digest: str
    unknown_preservation_digest: str
    prohibited_output_digest: str
    check_count: int
    citation_count: int
    unknown_count: int
    outcome: str
    adjudicated_at: datetime
    expires_at: datetime
    instance_state: str
    purpose: str
    canonical_digest: str
    knowledge_retrieved: bool = True
    model_context_available: bool = True
    model_invoked: bool = True
    protected_draft_available: bool = True
    model_draft_adjudicated: bool = True
    answer_generated: bool = False
    graph_updated: bool = False
    scheduled: bool = False
    workflow_continued: bool = False
    execution_authorized: bool = False
    deployment_approved: bool = False
    infrastructure_mutation_performed: bool = False
    reused: bool = False

    def __post_init__(self) -> None:
        later = (
            self.answer_generated,
            self.graph_updated,
            self.scheduled,
            self.workflow_continued,
            self.execution_authorized,
            self.deployment_approved,
            self.infrastructure_mutation_performed,
        )
        if (
            self.version != 1
            or self.instance_state != "protected_model_draft_adjudicated"
            or self.outcome
            not in {"adjudication-outcome.eligible", "adjudication-outcome.rejected"}
            or not all(
                (
                    self.knowledge_retrieved,
                    self.model_context_available,
                    self.model_invoked,
                    self.protected_draft_available,
                    self.model_draft_adjudicated,
                )
            )
            or any(later)
            or self.check_count < 1
            or self.citation_count < 0
            or self.unknown_count < 0
            or not 20 <= len(self.purpose.strip()) <= 1_000
            or self.adjudicated_at.tzinfo is None
            or self.expires_at.tzinfo is None
            or not self.adjudicated_at < self.expires_at
        ):
            raise ValueError("Protected draft-adjudication record is invalid")


@dataclass(frozen=True, slots=True)
class ProtectedDraftAdjudicationManifest:
    adjudication_id: str
    invocation_id: str
    context_id: str
    outcome: str
    check_count: int
    citation_count: int
    unknown_count: int
    report_digest: str
    check_set_digest: str
    citation_coverage_digest: str
    unknown_preservation_digest: str
    prohibited_output_digest: str
    adjudicated_at: datetime
    expires_at: datetime


@dataclass(frozen=True, slots=True)
class ProtectedDraftAdjudicationResult:
    record: ProtectedDraftAdjudicationRecord
    manifest: ProtectedDraftAdjudicationManifest

File: frontend/src/api/protectedDraftAdjudication.ts
import { apiFetch } from "./client";
import type { ProtectedModelInvocationResult } from "./protectedModelInvocation";

export type ProtectedDraftAdjudicationResult = {
  adjudication: {
    adjudication_id: string;
    schema_version: "atlas.protected-draft-adjudication.v1";
    version: 1;
    invocation_id: string;
    invocation_digest: string;
    context_id: string;
    organization_id: string;
    environment_id: string;
    adjudication_policy_id: string;
    adjudicator_id: string;
    outcome: string;
    canonical_digest: string;
    model_invoked: true;
    protected_draft_available: true;
    model_draft_adjudicated: true;
    answer_generated: false;
    graph_updated: false;
    scheduled: false;
    workflow_continued: false;
    execution_authorized: false;
    deployment_approved: false;
    infrastructure_mutation_performed: false;
  };
  manifest: {
    adjudication_id: string;
    invocation_id: string;
    context_id: string;
    outcome: string;
    check_count: number;
    citation_count: number;
    unknown_count: number;
    report_digest: string;
    check_set_digest: string;
    citation_coverage_digest: string;
    unknown_preservation_digest: string;
    prohibited_output_digest: string;
    adjudicated_at: string;
    expires_at: string;
  };
};

const forbidden = ["summary", "unknowns", "draft", "evidence", "check_codes", "protected_report_reference", "protected_report_digest", "consumer_subject_digest", "browser_session_binding_digest", "adjudication_authorization_digest", "prompt", "endpoint_url", "secret_reference", "tool_call"];

function isSafe(value: unknown): value is { data: ProtectedDraftAdjudicationResult } {
  if (!value || typeof value !== "object" || !("data" in value)) return false;
  const data = (value as { data?: unknown }).data;
  if (!data || typeof data !== "object") return false;
  const result = data as Record<string, unknown>;
  const adjudication = result.adjudication;
  const manifest = result.manifest;
  if (!adjudication || typeof adjudication !== "object" || !manifest || typeof manifest !== "object") return false;
  const record = adjudication as Record<string, unknown>;
  const safeManifest = manifest as Record<string, unknown>;
  return record.schema_version === "atlas.protected-draft-adjudication.v1" && record.model_draft_adjudicated === true && record.answer_generated === false && record.graph_updated === false && record.scheduled === false && record.workflow_continued === false && record.execution_authorized === false && record.deployment_approved === false && record.infrastructure_mutation_performed === false && typeof safeManifest.check_count === "number" && forbidden.every((field) => !(field in record) && !(field in safeManifest));
}

export async function createProtectedDraftAdjudication(input: { invocationResult: ProtectedModelInvocationResult; policyId: string; policyDigest: string }) {
  const { invocationResult, policyId, policyDigest } = input;
  const invocation = invocationResult.invocation;
  if (!invocation.model_invoked || !invocation.protected_draft_available || invocation.answer_generated || !/^[a-f0-9]{64}$/.test(policyDigest)) throw new Error("An exact protected model invocation is required");
  const response = await apiFetch(`/api/v1/ai/model-invocations/${encodeURIComponent(invocation.invocation_id)}/adjudications`, {
    method: "POST",
    headers: { Accept: "application/json", "Content-Type": "application/json", "Idempotency-Key": `protected-draft-adjudication.${crypto.randomUUID()}` },
    body: JSON.stringify({ schema_version: "atlas.protected-draft-adjudication-input.v1", invocation_digest: invocation.canonical_digest, adjudication_policy_id: policyId, adjudication_policy_digest: policyDigest, purpose: invocation.purpose, acknowledged_draft_is_untrusted: true, acknowledged_no_content_presentation: true, acknowledged_no_answer_or_operational_authority: true }),
  });
  if (!response.ok) throw new Error(`Protected draft adjudication failed with ${response.status}`);
  const payload: unknown = await response.json();
  if (!isSafe(payload)) throw new Error("Draft adjudication returned protected content or authority");
  return payload;
}
